### DISCUSSION TITLE
IMX378-190 General Comments

### DISCUSSION THREAD

**Discussion by StarGeezerPhil** [aaronwmorris/indi-allsky#568] [General]
First off - huge thank you @aaronwmorris for introducing support for the IMX sensors so quickly, really appreciate it.

Especially - also covering off the defective pixel correction (DPC) (Star-Eater Algorithm) so quickly too! 🔥

I'm testing with the IMX378-190, so just my notes so far for anyone else using this Waveshare module for the Pi...

- **Enable Camera**: Follow the [Waveshare instructions](https://www.waveshare.com/wiki/IMX378-190_12.3MP_Camera), or apply the following steps to get the camera working:
`sudo nano /boot/config.txt`
Then add the following to the end of the file:
`dtoverlay=imx378`

- **Colour balance** in INDI-AllSky - I've found this required quite significant  changes (not tested with auto-white-balance), but these settings seem to be close to optimal for me: 0.65, 0.55, 1.10 (R:G:B)

- **Resolution** is too high for the **timelapse** video creation, however I'd rather keep my high resolution individual images than resize them all. Is it possible to add a resize option for just ffmpeg anyone?

- **Gain**; darks script indicates maximum value is 16.

- **Noise**; there is A LOT of noise. Capture darks! However, there is currently no sensor temperature captured in INDI-AllSky, so they won't be temperature calibrated - but even so, it makes a HUGE improvement to the image quality.

- **Disable LEDs**: Not IMX378-specific, but Pi users - I'd recommend disabling the LEDs to avoid introducing unwanted light into your exposures:
`sudo nano /boot/config.txt`
Then, add the following to the end of the file (for Pi 4, presumably 3 too?) - broken up as github markdown turns comments into titles:

**Disable the PWR LED**
```
dtparam=pwr_led_trigger=none
dtparam=pwr_led_activelow=off
```
**Disable the Activity LED**
```
dtparam=act_led_trigger=none
dtparam=act_led_activelow=off
```
**Disable ethernet port LEDs**
```
dtparam=eth_led0=4
dtparam=eth_led1=4
```
(Sadly, it seems impossible to disable the LED on the Pi POE-Hat - if anyone knows how, please let me know)

**Comment by aaronwmorris** [comment]
Thanks for this!

I will try get some of my wiki pages updated with this info.

* Resolution - I will think about what it would take to implement resizing the images just for ffmpeg.
* Gain - I assume the gain options are the same for the IMX477, but that could be wrong.  If it is wrong, I will have to create a different profile for the IMX378.
* Disable LEDs - I have a wiki page for this, but I think it would be helpful to add this option as an alternative.  https://github.com/aaronwmorris/indi-allsky/wiki/Disable-LEDs
    * I did not know it was possible to disable the NIC leds, that is really cool

**Comment by StarGeezerPhil** [comment]
No trouble - anything else I find I'll post.

Not managed to find anything to disable the big red LED on the PoE HAT sadly.

Yeah, if there's anything you can do for ffmpeg it'd be great mate - have a lot of captures that I simply can't timelapse at the moment. 

In the JQ Allsky he has a ffmpeg commandline option, which would allow me to do some testing for you, but suspect it'd be something like I put in the other bug thread, e.g. -vf scale=-1:2304 (assuming the scaling option can work with either X or Y - it might be limited to just X resolution).

**Comment by aaronwmorris** [comment]
You could dab a bit of caulk for a non-permanent fix.  I tested some other SoCs and they had some LEDs that could not be turned off.  I just broke the LEDs off with needle nose pliers.  FYI, SMT LEDs come off much easier than I thought they would.

**Comment by StarGeezerPhil** [comment]
I've just moved it into a solid opaque case - so it's less of an issue now, but would be good to permanently disable it. Waiting on a new dome to fit and looking at some flocking this time around to reduce reflections in the dome.

**Comment by aaronwmorris** [comment]
I am going to play with adding an option to scale the vertical resolution

**Comment by StarGeezerPhil** [comment]
Ping me if you get a working test - be happy (and keen) to test it out 👍

**Comment by aaronwmorris** [comment]
I have been testing adding the option and I just happened to test trying to encode the original resolution of the images and ffmpeg allowed it...  I wonder if this was just a limitation of ffmpeg in Buster.

![Screen Shot 2022-08-29 at 9 42 13 PM](https://user-images.githubusercontent.com/17464290/187329162-36b4e2bf-aa0c-425d-a8e2-cb2127b8916a.png)

**Comment by aaronwmorris** [comment]
Confirmed.  I see in my log ffmpeg used profile level 6.0.

Newer versions of ffmpeg support H264 level 6.0+ which supports higher resolutions up to 8192×4320.

https://en.wikipedia.org/wiki/Advanced_Video_Coding

**Comment by StarGeezerPhil** [comment]
That IS exciting news! Going to need a bigger SD card 🤣

Look forward to the updates Aaron 🥳👍

**Comment by aaronwmorris** [comment]
I have merged the new config option #335  to perform the vertical scaling.

I had originally intended on adding more options like a standard 1080 (1k) and 2160 (4k) option, but it is not that easy.  Whatever scaling option you choose has to result in a horizontal and vertical resolution that is divisible by 2 and this does not always happen.

I can always add more options, but each option will likely be camera specific.

**Comment by aaronwmorris** [comment]
Hah... my Raspi4 4GB ran out of memory trying to build a full resolution timelapse video

> [290258.454992] Out of memory: Killed process 148816 (ffmpeg) total-vm:3640732kB, anon-rss:2697684kB, file-rss:1888kB, shmem-rss:0kB, UID:1000 pgtables:5940kB oom_score_adj:0

**Comment by StarGeezerPhil** [comment]
Hi Aaron, sorry life took over for the last month and the cover for my V2 skycam got lost in the post... finally back on it again 🥳

Hopefully fully waterproof, with some ventilation, IMX378 and flocking beneath the dome this time around, so 🤞


I'm having some trouble with the image and video generation.

Video generation - seems to be the bitrate more than anything else that affects it. Was hoping to generate higher quality video at 2304px, 75%, or 1520px, but if I increase bitrate beyond 4000k the video seems to fail. When I reduced to the 4000k bitrate my timelapse is now generating.

Images - getting neither startrail or keogram generated from the IMX378 images from last night's first test...

...could you add the troubleshooting cmd line options into the wiki?
(I know there's an issue logged somewhere with the details but it's difficult to find).

Also, just general UX suggestion (though, I guess once we're each setup and optimised we rarely need to go back into config):
- maybe separate the options into tabs like "Environment", "Imaging", "Processing", "Output", "FTP/Upload" or something similar. The config is GREAT - just that there are so many options now I'm often scrolling up/down the list and there's few visual queues to stop me in the right place. Just some headings with something visual (e.g. bootstrap icon) would help the end-user quickly navigate to the right settings. 

No complaint - just an idea/feedback 👍

**Reply by aaronwmorris** [reply]
Merged #654 to separate config sections.

**Comment by aaronwmorris** [comment]
> Video generation - seems to be the bitrate more than anything else that affects it. Was hoping to generate higher quality video at 2304px, 75%, or 1520px, but if I increase bitrate beyond 4000k the video seems to fail. When I reduced to the 4000k bitrate my timelapse is now generating.

Check the system log with `sudo dmesg` and check for out of memory messages.  That is one of the most common reasons for ffmpeg to fail other than inputting images of different sizes.

> Images - getting neither startrail or keogram generated from the IMX378 images from last night's first test...

If you changed any of the image scaling options, this can happen.  All of the images must be exactly the same dimensions in order for the keogram or startrails to generate.

> maybe separate the options into tabs like "Environment", "Imaging", "Processing", "Output", "FTP/Upload" or something similar

The size of the config page has grown considerably and I will be splitting it into multiple tabs soon.

**Comment by StarGeezerPhil** [comment]
Hi Aaron, what was your conclusion about gain for your sensor.

I thought that gain went up to 16 as I commented above, however last night it appeared there was no difference for gain set between 5-16. (Unless I didn't apply the config correctly).

I'll continue digging.

**Comment by aaronwmorris** [comment]
In order to change the gain, you have to fully restart indi-allsky.  The reload option is not sufficient.  (I probably need to note this somewhere)

There is not a lot of information about the IMX378 out there.  I cannot even find a datasheet.  I found the datasheet for the IMX377 which says gain is up to +27dB in 0.1 increments.  But this does not mean anything with libcamera.  ZWO allows you to set gain in 0.1 increments... 270 means 27dB.  QHY supports the whole numbers so 27 would mean 27.  Anything above these would go into digital gain (instead of analog).

I may have to hunt through the libcamera source code.

**Comment by StarGeezerPhil** [comment]
Hi @aaronwmorris,

I've just swapped out the sealant, remounted the camera higher (190° lens - originally capturing a lot of the box/silicone) and I'm trying to collect more darks.

(Back to one of our earlier conversations about gain...)

The dark script reports gain 16 as the max for the IMX378 - reported as below:
```
2023-03-15 15:05:56,301 [ERROR] MainProcess darks._initialize() #269: CCD night gain above maximum, changing to 16
2023-03-15 15:05:59,305 [ERROR] MainProcess darks._initialize() #278: CCD moon mode gain above maximum, changing to 16
```

**Reply by aaronwmorris** [reply]
I can increase the max gain value in the config.  It is difficult to find hard data on the gains for the cameras.  I think the max analog  gain is 22.25 or 22.26.  Do you have any reference on this?

**Reply by StarGeezerPhil** [reply]
No idea, sorry Aaron. I'm just happy as long as the darks match what it's running with - I had tried setting high gain >150, but I'll revert to using 16 and see how it goes.

With the >150 values set do you happen to know if would have done the same during capture (reset itself to 16 and use those darks)?

**Reply by aaronwmorris** [reply]
I have bumped the max gain for imx477 and imx378 to 22 in #639

**Reply by aaronwmorris** [reply]
The darks are stored with the gain value as they were taken and will only be used if the gain matches.  If you change your gain,  you will have to retake the darks.

I had to implement the safety mechanism on the gains to prevent incorrect settings.  There are so many variations between cameras, there was no way to provide logical defaults without placing soft limits.

**Reply by StarGeezerPhil** [reply]
Thanks Aaron, I'll update and take a set at 22 while I'm doing this 👍

Sadly, struggling to get the sensor below 20°C - should have done it before I attached the dome 🤣

**Reply by StarGeezerPhil** [reply]
Just a feedback idea re: taking the darks:

Could you add the sensor temperature at the end of each exposure? Reason: if you're trying to achieve darks at a specific temperature and you're taking a long exposure (e.g. 60s) you have to wait until the very end (e.g. 11 minutes) to discover the temperature of the darks.

Maybe an additional logging/output line:

```
2023-03-15 18:25:19,676 [INFO] MainProcess libcamera.setCcdExposure() #192: image command: libcamera-still --immediate --nopreview --raw --denoise off --awbgains 1,1 --gain 8 --shutter 40000000 --metadata /tmp/tmp_j2od5ia.json --metadata-format json --output /tmp/tmp5edjyldb.dng
2023-03-15 18:26:01,143 [INFO] MainProcess darks._take_exposures() #660: Exposure received in 41.4680 s
2023-03-15 18:26:01,536 [INFO] MainProcess darks._take_exposures() #682: Image average adu: 259.00
2023-03-15 18:26:01,536 [INFO] MainProcess darks._take_exposures() #692: Sensor temperature: 20.00°C
```

**Reply by aaronwmorris** [reply]
Merged #641

**Reply by StarGeezerPhil** [reply]
You're on fire dude 🔥

**Reply by StarGeezerPhil** [reply]
@aaronwmorris just experimenting with binning for the first time:

Night Mode > Night Bin Mode: 2

I was expecting the resulting images to be a lower resolution, however pre- and post- settings (with full reboot to be sure) are both 4056x3040 resolution... is this correct?

**Reply by aaronwmorris** [reply]
The binmode setting in the config does not change the binning for libcamera.  You have to manually specify that in the extra arguments.  You should still change the binmode setting in the config so the darks are properly marked.

https://github.com/aaronwmorris/indi-allsky/wiki/libcamera-enablement

**Reply by StarGeezerPhil** [reply]
I was going to ask about the darks (I spent around 7 hours collecting darks yesterday)... do we actually need separate darks for different bin settings? Wouldn't they all run from the full frame... then get binned subsequently?

I'll take a look at the above - thank you. Didn't think of looking there in the wiki - might be worth adding a page for binning, referring to this for libcamera.

**Reply by StarGeezerPhil** [reply]
@aaronwmorris might be an issue (or I've added the command line options incorrectly):

```
Mar 16 19:16:51 SkyCam [INFO] MainProcess allsky.detectNight() #1558: Sun altitude: -10:33:40.3
Mar 16 19:16:50 SkyCam [INFO] MainProcess allsky._startImageWorker() #765: Starting ImageWorker process 3
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception: 
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception: 
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception: cv2.error: OpenCV(4.6.0) /io/opencv/modules/core/src/arithm.cpp:647: error: (-209:Sizes of input arguments do not match) The operation is neither 'array op array' (where arrays have the same size and the same number of channels), nor 'array op scalar', nor 'scalar op array' in function 'arithm_op'
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:     data_calibrated = cv2.subtract(data, master_dark)
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:   File "/home/StarGeezer/indi-allsky/indi_allsky/image.py", line 1864, in _calibrate
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:     calibrated_data = self._calibrate(i_ref['hdulist'][0].data, i_ref['exposure'], i_ref['camera_id'], i_ref['image_bitpix'])
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:   File "/home/StarGeezer/indi-allsky/indi_allsky/image.py", line 1727, in calibrate
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:     self.image_processor.calibrate()
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:   File "/home/StarGeezer/indi-allsky/indi_allsky/image.py", line 294, in processImage
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:     self.processImage(i_dict)
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:   File "/home/StarGeezer/indi-allsky/indi_allsky/image.py", line 223, in saferun
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:     self.saferun()
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception:   File "/home/StarGeezer/indi-allsky/indi_allsky/image.py", line 194, in run
Mar 16 19:16:50 SkyCam [ERROR] MainProcess allsky._startImageWorker() #758: Image worker exception: Traceback (most recent call last):
Mar 16 19:16:50 SkyCam [INFO] MainProcess allsky.run() #1033: Exposure state: READY
Mar 16 19:16:50 SkyCam [INFO] MainProcess allsky.run() #1032: Camera last ready: 4.1s
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image._calibrate() #1851: Matched dark: /var/www/html/allsky/images/darks/dark_ccd2_16bit_1s_gain22_bin2_19c_20230315_220216.fit
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image._calibrate() #1835: Matched bad pixel map: /var/www/html/allsky/images/darks/bpm_ccd2_16bit_1s_gain22_bin2_19c_20230315_220216.fit
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image._calibrate() #1784: Searching for dark frame: gain 22, exposure >= 0.0, temp >= 19.0c
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image._calibrate() #1737: Searching for bad pixel map: gain 22, exposure >= 0.0, temp >= 19.0c
Mar 16 19:16:46 SkyCam [WARNING] ImageWorker002 image._detectBitDepth() #1709: Updated default bit depth: 12
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image._detectBitDepth() #1681: Image max value: 3454
Mar 16 19:16:46 SkyCam [INFO] ImageWorker002 image.add() #1628: Image bits: 16, cfa: BGGR
Mar 16 19:16:46 SkyCam [INFO] MainProcess allsky.run() #1174: Exposure received in 1.0844 s (1.0834)
Mar 16 19:16:45 SkyCam [INFO] MainProcess allsky.run() #1243: Total time since last exposure 45.0106 s
Mar 16 19:16:45 SkyCam [INFO] MainProcess libcamera.setCcdExposure() #192: image command: libcamera-still --immediate --nopreview --raw --denoise off --awbgains 1,1 --gain 22 --shutter 1000 --metadata /tmp/tmpqfet9517.json --metadata-format json --mode 2028:1520 --output /tmp/tmp9iiruj3l.dng
```

**Reply by aaronwmorris** [reply]
Binning would require separate darks.  Libcamera makes things more complex by the different cropping modes and packed vs unpacked modes.

**Reply by aaronwmorris** [reply]
I do need to put extra checks to ensure the darks have the same dimensions as the images being generated.

**Reply by StarGeezerPhil** [reply]
I'm guessing that reply is to the first comment? Or is it because I collected the darks without applying the command-line parameter?

**Reply by StarGeezerPhil** [reply]
Don't tell me I wasted ~3 1/2 hours 😭🤣

**Reply by StarGeezerPhil** [reply]
So, collecting darks when binning with libcamera:

1. Set the commandline parameters in settings > libcamera extra options (e.g. --mode 2028:1520)
2. Restart system
3. Follow standard darks procedure

No trouble, I'd just tried to collect darks for every gain setting I'm likely to use - with a bin: 2 variant for each. Hoping to mount on the roof ASAP, so was just trying to cover all bases.

**Reply by StarGeezerPhil** [reply]
Sorry - follow-up question, is there a way I can flush ONLY the bin: 2 darks?

**Reply by aaronwmorris** [reply]
Sorry :-(

You can just delete the individual files manually.

**Reply by aaronwmorris** [reply]
Merged #644 to ensure dimensions match.

If I could go back, I would probably add the dimensions to the SQL table for darks, but since indi-allsky was designed around indi where the image size is almost always associated with the binmode of the camera.  libcamera just does not have a standard for setting the binmode

**Reply by StarGeezerPhil** [reply]
No worries 👍

Deleting the individual files and indi-allsky will detect they're missing and clear the dB entries okay? (I was worried it'd cause errors if it thought it had them and then they don't exist)

I think I mentioned this a while ago in passing (and only low priority / nice-to-have): would it be possible to add an image overlay that could blank out areas of the image (essentially a reverse of the detection mask).

E.g. Example usages:
- Tidy up the image
- Privacy, e.g. features a bedroom window of a neighbour for instance
- Branding/Camera/Website info

Example:
RAW image:
![image](https://user-images.githubusercontent.com/94135145/225741890-b882d40c-54a0-4b19-b648-5469caf01846.png)

Overlay:
![image](https://user-images.githubusercontent.com/94135145/225741450-89c6b436-e822-434d-ade9-5162b566fa7a.png)

Resulting Image:
![image](https://user-images.githubusercontent.com/94135145/225741480-ae05fe34-ba5d-4620-a906-1bd7c7b4a66f.png)

**Reply by aaronwmorris** [reply]
In the system section, if you "Validate" the database entries, it will remove any dark frame database entries without files on the filesystem.

**Reply by aaronwmorris** [reply]
Merged #645 with support for a privacy mask.

**Reply by aaronwmorris** [reply]
I should have read this more closely.  I'll probably completely change this to support adding a logo.

**Reply by aaronwmorris** [reply]
I'm getting closer....

![fake_jupiter](https://user-images.githubusercontent.com/17464290/225794906-1585a606-674e-4a06-9998-2a0a4973254e.jpg)

**Reply by StarGeezerPhil** [reply]
Thanks Aaron. 

The priority for me is to simply tidy up the area around the lens (out of shot) - like the above example, it's often got lots of light bleeding/internal reflections - I just added your avatar and some text as an example of how the overlay could also be used to add graphical content as well.

But, yes - you could have multiple layers. One layer to tidy the image, another one (or more) to add different graphic elements like a logo, etc.

**Reply by aaronwmorris** [reply]
Merged #646 .  Added a logo overlay option (replaces the privacy mask).

I managed to get this to support full alpha transparency.  The logo image has to have an alpha channel, but allows for things like shadows on the logo.  The nice thing about the alpha channel, you do not need to dedicate a color like white or black for the transparency, the logo layer can be full color.

![overlay_test](https://user-images.githubusercontent.com/17464290/225935686-7d43e79b-f866-4ca7-aa3a-2c23507b6402.jpg)

**Reply by StarGeezerPhil** [reply]
Awesome work, thanks Aaron 💖👍

Really appreciate that - has tidied up the outer image to no end.
